### ISSUE TITLE
Feature request: Reduce learning rate on plateau

### ISSUE THREAD

**Issue by doktorneergaard** [tensorflow/tensorflow#15983]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes. But applies to stock examples as well.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04.
- **TensorFlow installed from (source or binary)**:
Binary.
- **TensorFlow version (use command below)**:
v1.4.0-19-ga52c8d9 1.4.1
- **Python version**: 
Python 3.5.4
- **Bazel version (if compiling from source)**:
N/A
- **GCC/Compiler version (if compiling from source)**:
N/A
- **CUDA/cuDNN version**:
CUDA: V8.0.61
cuDNN: 6.0.21
- **GPU model and memory**:
GTX 1080Ti 11GB running driver version 384.98
- **Exact command to reproduce**:
N/A. However, I am using the Experiment, Estimator and Dataset APIs in order to do training.

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.
I would like to reduce the learning rate during training. However, I do not want to treat this as another tunable hyperparameter, so I would like this to be based on performance plateauing. In Keras, it is easy to implement learning rate reduction by monitoring the validation loss using the ReduceLROnPlateau callback function, but in TensorFlow this does not seem to be the case/easy. I certainly haven't found any implementation of this, so I propose this as a feature request.

Feel free to close this if this is not the correct forum.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by shivaniag** [comment]
@martinwicke does it seems likes a feature we would want to have or there is any workaround for it?

**Comment by martinwicke** [comment]
The answer depends on how you are using TensorFlow. 

If you are writing plain low-level TensorFlow code and your own training loop (e.g., using MonitoredTrainingSessions interleaved with validation, this is very straightforward. The learning rate is a parameter to the optimizers, which is a Tensor. The Tensor could be computed any way you like to, including any sort of plateau detection (presubmably, by comparing two rolling averages of the loss), and if it triggers, reducing the learning rate. We may not have a function that does exactly what you want, but it is easy to write. 

If you are using an Estimator, you do not have easy access to the validation loss inside the training loop. This is because in distributed training, evaluation is performed on a separate worker, and in an isolated environment, while training continues on the other workers. If you want to implement learning rate decay based on evaluation loss there, you would write a hook which reads the loss from the evaluation event file, does whatever logic it needs to, and assigns a new learning rate to a training variable, which in turn is used as the learning rate in the optimizer. This is not hard to do, but it is not done.

Of course, you can also use tf.keras if that works for you, that's why it's there.

**Comment by doktorneergaard** [comment]
@martinwicke Thanks for the nice response. I am using the Experiment/Estimator combo, though I am not planning to do distributed training. 

However, wouldn't it be possible to create a loop over a training and evaluation phase using Estimator alone, grab the loss from the evaluate() call, determine learning rate decay and then reset the learning rate for the subsequent training/evaluation phases?

Something like this pseudo:
```
# Define model
model = tf.estimator.Estimator(...)

# Cycle over epochs
for cycle in num_cycles:
    model.train(...)

    # Grab loss
    loss = model.evaluate(...)

    # Determine if hit plateau:
    if hit_plateau:
        reduce_learning_rate()
```
Where of course the `reduce_learning_rate()` function would impute the new learning rate value to the learning rate tensor. Would that be a possible workaround?

**Comment by martinwicke** [comment]
TFRecords and Keras is possible, but not necessarily simple.

The pseudocode you describe is possible. You can add learning rate to the params arg (which is passed to the model_fn). 

Then, make a new estimator in each iteration of the loop. The Estimator will read the last state from disk anyway every time you call one of its methods. 

It would look like this:

```
def model_fn(..., params):
    # Do something with params['learning_rate']
    ...

lr = INITIAL_LEARNING_RATE

# Cycle over epochs
for cycle in num_cycles:
    # Define model
    model = tf.estimator.Estimator(..., params={'learning_rate': lr})

    model.train(...)

    # Grab loss
    loss = model.evaluate(...)

    # Determine if hit plateau:
    if hit_plateau:
        lr = reduce_learning_rate(lr)
```

Note that constructing the Estimator is effectively free. Calling train is a loop is not something I recommend, but for local training it's ok, especially if you only call it once per epoch.

**Comment by sarbiv** [comment]
Hey,
Is there any intention to implement ReduceLROnPlateau that is integrated with tf.estimators? or your recommendation is to use tf.keras in order to achieve that behavior?
Thanks

**Comment by Jasonnor** [comment]
+1
